datumaro/components/extractor.py

```diff
@@ -673,7 +673,8 @@ def __call__(self, path, **extra_params):
     @classmethod
     def _find_sources_recursive(cls, path, ext, extractor_name,
             filename='*', dirname='', file_filter=None, max_depth=3):
-        if path.endswith(ext) and osp.isfile(path):
+        if (path.endswith(ext) and osp.isfile(path)) or \
+            (ext == '' and osp.isdir(path) and not dirname == '' and dirname in path):
```

Comment:
In my case your suggested check with `endwith` is not working, because for cityscapes we got for example  `imgsFine/leftImg8bit` as `dirname` and `Cityscapes/only_val/imgsFine/leftImg8bit/val` as `path`. But ok, will think how to make check better

Reply:
`osp.normpath(dirname) in osp.abspath(path)` should be better check

Reply:
This check still has the problem with `a/b` in `/c/a/bd/`.

Reply:
Updated, made more tricky check

Reply:
I think, we can try `osp.isdir(path) and  '%s%s%s' % (os.sep, osp.normpath(dirname), os.sep) in osp.abspath(path) + os.sep`

Reply:
This looks simple and correct! updated